tensorflow/contrib/tensorrt/convert/convert_graph.cc

```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+    }
+  }
+  status = node_builder.Attr("input_shapes", input_shapes)
+               .Attr("output_shapes", out_shapes)
+               .Attr("static_engine",
+                     info.engine_type == EngineInfo::EngineType::TRTStatic)
+               .Attr("segment_funcdef_name",
+                     StrCat(info.engine_name, "_native_segment"))
+               .Attr("serialized_segment", segment_string)
+               .Attr("calibration_data", "")
+               .Attr("max_cached_engines_count", info.maximum_cached_engines)
+               .Attr("cached_engine_batches", {max_batch_size})
+               .Attr("workspace_size_bytes", info.max_workspace_size_bytes)
+               .Attr("precision_mode", prec_string)
+               .Attr("OutT", out_types)
+               .Finalize(trtNode);
+  if (!status.ok()) {
+    LOG(ERROR) << "Node construction failed with" << status;
+    return status;
+  }
+  VLOG(1) << "Adding TRTEngine " << info.engine_name << " to graph";
+  engine_node = graph->AddNode(*trtNode, &status);
+  if (!status.ok()) {
+    LOG(ERROR) << "Adding node failed " << status;
+    return status;
+  }
+
+  for (auto& conn : info.connections) {
+    if (conn.is_input_edge) continue;
+    VLOG(1) << " Updating DBG " << engine_node->name() << " out_port "
+            << conn.port_number << " out_id " << conn.outside_id
+            << " name=" << conn.outside_node_name;
+    auto dst_node = graph->FindNodeId(conn.outside_id);
+    if (!dst_node) {  // node removed skip.
+      continue;
+    }
+    VLOG(1) << "Updating " << engine_node->name() << ":" << conn.port_number
+            << " to " << dst_node->name() << ":" << conn.outside_port;
+    status = graph->UpdateEdge(engine_node, conn.port_number, dst_node,
+                               conn.outside_port);
+    if (!status.ok()) {
+      LOG(ERROR) << "Edge update failed " << engine_node->name() << ":"
+                 << conn.port_number << " -> " << dst_node->name() << ":"
+                 << conn.outside_port << " status= " << status;
+    }
+  }
+  return status;
+}
+
+// Function to construct a funcdef from the segment and add it to the graph.
+tensorflow::Status RegisterSegmentFunctionToFunctionLibrary(
+    tensorflow::Graph* graph, const tensorflow::GraphDef& segment,
+    const string& name) {
+  tensorflow::Graph sgraph(graph->flib_def());
+  tensorflow::GraphConstructorOptions gcopts;
+  TF_RETURN_IF_ERROR(
+      tensorflow::ConvertGraphDefToGraph(gcopts, segment, &sgraph));
+  std::map<string, tensorflow::Node*> io_nodes;
+  int num_inputs = 0;
+  for (auto n : sgraph.op_nodes()) {
+    if (tensorflow::str_util::StartsWith(n->name(), "InputPH_")) {
+      num_inputs++;
+      io_nodes.insert({n->name(), n});
+    } else if (tensorflow::str_util::StartsWith(n->name(), "OutputPH_")) {
+      io_nodes.insert({n->name(), n});
+    }
+  }
+
+  for (int i = 0; i < num_inputs; ++i) {
+    auto name = StrCat("InputPH_", i);
```

Comment:
I don't quite understand: why do the conversion from InputPH_ to _Arg here? Why not directly construct the _Arg nodes from the very beginning?